File: examples/microgrid/simulators/battery.py
# -*- coding: utf-8 -*-
"""Battery energy storage system — external simulator for energysim.

Models a Li-ion battery (e.g. Tesla Powerwall) with Euler-integrated
state-of-charge dynamics.  Capacity 13.5 kWh, max power 5 kW.
"""

import math
from energysim.base import SimulatorAdapter


class external_simulator(SimulatorAdapter):
    """Li-ion battery with Euler-integrated SoC dynamics."""

    def __init__(self, inputs=None, outputs=None, **kwargs):
        self.inputs = inputs or []
        self.outputs = outputs or []

        # State
        self.SoC = 0.5

        # Parameters
        self.capacity = 0.0135      # MWh (13.5 kWh)
        self.max_power = 0.005      # MW  (5 kW)
        self.efficiency = 0.95

        # I/O variables
        self.P_cmd = 0.0            # MW (positive = charge)
        self.P_actual = 0.0         # MW

    def init(self):
        pass

    def set_value(self, parameters, values):
        for p, v in zip(parameters, values):
            if p == 'P_cmd':
                self.P_cmd = v

    def get_value(self, parameters, time):
        result = []
        for p in parameters:
            if p == 'SoC':
                result.append(self.SoC)
            elif p == 'P_actual':
                result.append(self.P_actual)
            else:
                result.append(0.0)
        return result

    def step(self, time):
        dt = self.step_size
        sqrt_eff = math.sqrt(self.efficiency)

        # Clip command to rated power
        P = max(-self.max_power, min(self.max_power, self.P_cmd))

        # Compute SoC change  (capacity in MWh → MWs = capacity * 3600)
        if P >= 0:
            dSoC = P * sqrt_eff * dt / (self.capacity * 3600)
        else:
            dSoC = P / sqrt_eff * dt / (self.capacity * 3600)

        new_SoC = self.SoC + dSoC

        # Enforce SoC limits
        if new_SoC > 1.0:
            self.SoC = 1.0
            self.P_actual = 0.0
        elif new_SoC < 0.0:
            self.SoC = 0.0
            self.P_actual = 0.0
        else:
            self.SoC = new_SoC
            self.P_actual = P

    def cleanup(self):
        pass

File: examples/matlab_cosim/models/battery.m
function [SoC, P_actual] = battery(P_cmd, time)
% BATTERY  Simple Li-ion battery model with persistent state.
%
%   [SoC, P_actual] = battery(P_cmd, time)
%
%   Inputs
%       P_cmd   - Power command [kW].  Positive = discharge, negative = charge.
%       time    - Current simulation time [s]  (appended by energysim)
%
%   Outputs
%       SoC      - State of charge [0 .. 1]
%       P_actual - Actual power delivered/absorbed [kW]
%                  (may differ from P_cmd when SoC limits are hit)
%
%   The model uses persistent variables to keep track of the SoC and
%   the previous time-step.  energysim calls the function once per
%   macro time-step; the function derives dt from consecutive calls.
%
%   Parameters (edit inside this file):
%       E_cap    = 13.5 kWh   (Tesla Powerwall-sized)
%       eta_ch   = 0.95       charge efficiency
%       eta_dis  = 0.95       discharge efficiency
%       P_max    = 5.0 kW     max charge/discharge power
%       SoC_min  = 0.10
%       SoC_max  = 0.90

    % --- Parameters ---
    E_cap   = 13.5;     % usable capacity      [kWh]
    eta_ch  = 0.95;     % charge efficiency     [-]
    eta_dis = 0.95;     % discharge efficiency  [-]
    P_max   = 5.0;      % max power             [kW]
    SoC_min = 0.10;     % minimum SoC           [-]
    SoC_max = 0.90;     % maximum SoC           [-]

    % --- Persistent state ---
    persistent soc t_prev
    if isempty(soc)
        soc    = 0.50;   % initial SoC
        t_prev = time;
    end

    % --- Time step ---
    dt = (time - t_prev) / 3600.0;   % seconds -> hours
    t_prev = time;
    if dt <= 0
        % First call or zero-length step
        SoC      = soc;
        P_actual = 0.0;
        return;
    end

    % --- Clamp command to rated power ---
    P_actual = max(-P_max, min(P_cmd, P_max));

    % --- Energy balance ---
    if P_actual >= 0
        % Discharging
        dE = P_actual * dt / eta_dis;   % energy removed from battery
    else
        % Charging
        dE = P_actual * dt * eta_ch;    % energy added (negative P -> negative dE if charging adds)
        dE = P_actual * dt * eta_ch;    % dE < 0  (energy stored)
    end

    soc_new = soc - dE / E_cap;

    % --- SoC limits ---
    if soc_new > SoC_max
        % Reduce charge to stay at SoC_max
        dE_avail = (soc - SoC_max) * E_cap;
        if P_actual < 0
            P_actual = -dE_avail / (dt * eta_ch);
        end
        soc_new = SoC_max;
    elseif soc_new < SoC_min
        % Reduce discharge to stay at SoC_min
        dE_avail = (soc - SoC_min) * E_cap;
        if P_actual > 0
            P_actual = dE_avail * eta_dis / dt;
        end
        soc_new = SoC_min;
    end

    soc = soc_new;
    SoC      = soc;
    P_actual = P_actual;
end
